### PR TITLE
Replace small matrix solver with vectorized Levinson recursion

### LINKED ISSUE #87: switch chi2 small matrix solver to eigen backend
We could probably increase the robustness of our chi2 backend if we used a high-quality linear solver like Eigen (https://libeigen.gitlab.io/eigen/docs-nightly/group__TutorialLinearAlgebra.html). Our current custom implementation is fast, but probably a little too eager to declare a matrix singular. We originally tried LAPACK, but that had high overhead for the small matrices that appear in LS. Eigen's solver probably has less overhead, since it doesn't involve a call to a shared library.

x-ref #85

### PR DESCRIPTION
**Pull request by kkrutkowski**
Resolves #87 

Draft of a PR replacing the small matrix solver with faster and generally more accurate (at least for small systems) [Levinson recursion](https://en.wikipedia.org/wiki/Levinson_recursion).

Important default behavior change - currently, NiftyLS throws an error after encountering first singular frequency. Changes result in replacing the behavior with returning NaN powers silently and returning an error only if the entire grid is filled with NaNs - generally should significantly increase the package’s usability for high-nterms applications (like fitting periods to separated eclipsing binaries or heartbeat stars), in which singularities are a norm rather than an exception.

Currently, the NaN threshold is set to 1e11 for an upper boundary of the conditionality of the problem, which is extremely conservative and might require adjustment before acceptance (main reason why this PR is set as draft). Of course, I am open to changes and/or direct rejection of this PR.

Additional note (as this PR might be a good moment to add it) - the “Fast χ^2 Method” section in the readme might require correction, as the multiharmonic generalization of the Scargle periodogram (I am not too sure if a multiharmonic version of Lomb’s periodogram is commonly used in any application, by the way) did exist for some time before Palmer. A. Schwartzenberg-Czerny proposed a mathematically equivalent algorithm in his 1996 paper [“Fast and Statistically Optimal Period Search in Uneven Sampled Observations”](https://iopscience.iop.org/article/10.1086/309985/fulltext/).

I am almost certain that D. Palmer’s contribution was limited to applying NuFFT to the extension and not developing it, although I would be grateful if anyone could correct me on that as well.

### CODE REVIEW

File: src/nifty_ls/chi2_helpers.hpp [hunk 1]
```diff
@@ -28,43 +31,40 @@ using namespace nb::literals;
...
-   Scalar *w2s,                           // (Nbatch, 1)
-   Scalar *norm,                          // (Nbatch, 1)
-   Scalar *yws,                           // (Nbatch, 1)
-   Scalar *Sw,                            // (Nbatch, nSW, Nf)
-   Scalar *Cw,                            // (Nbatch, nSW, Nf)
-   Scalar *Syw,                           // (Nbatch, nSY, Nf)
-   Scalar *Cyw,                           // (Nbatch, nSY, Nf)
-   const nifty_arr_1d<const Scalar> &t,   // (N_d)
-   const nifty_arr_2d<const Scalar> &y,   // (N_batch, N_d)
-   const nifty_arr_2d<const Scalar> &dy,  // (N_batch, N_d)
```
Comment (lgarrison) [inline src/nifty_ls/chi2_helpers.hpp:67]: Can you restore these comments (assuming they're still accurate)?

File: src/nifty_ls/chi2_helpers.hpp [hunk 2]
```diff
@@ -28,43 +31,40 @@ using namespace nb::literals;
...
-    SUCCESS  = 0,
-    SINGULAR = 1
-};
+#if defined(__AVX512F__)
+#define SOLVER_VEC_BYTES 64
+#elif defined(__AVX__)
+#define SOLVER_VEC_BYTES 32
+#else
+#define SOLVER_VEC_BYTES 16
+#endif
```
Comment (lgarrison) [inline src/nifty_ls/chi2_helpers.hpp:40]: The wheels target the default manylinux arch, so they won't get any benefit from AVX or AVX-512. But I think we can probably bump the target to at least support AVX, seems not too restrictive...

File: src/nifty_ls/chi2_helpers.hpp [hunk 3]
```diff
@@ -210,89 +210,165 @@ void compute_t_raw(
...
+template <typename Scalar>
+SOLVER_VEC<Scalar> solver_vec_sqrt(const SOLVER_VEC<Scalar> value) {
+    SOLVER_VEC<Scalar> result{};
+    for (size_t lane = 0; lane < SOLVER_VEC_LEN<Scalar>; ++lane) {
+        result[lane] = std::sqrt(value[lane]);
+    }
+    return result;
+}
+
+inline bool solver_is_nonfinite(const double value) {
```
Comment (lgarrison) [inline src/nifty_ls/chi2_helpers.hpp:231]: Let's drop a comment here explaining why we have these functions (because we're compiling with `-ffast-math`).

File: src/nifty_ls/chi2_helpers.hpp [hunk 4]
```diff
@@ -210,89 +210,165 @@ void compute_t_raw(
...
+            Ai[i]                        = Api[i] + gamma_r * api + gamma_i * apr;
         }
 
-        // Backward substitution: Solve U * X = Y
-        for (size_t i = n; i-- > 0;) {
-            for (size_t j = i + 1; j < n; ++j) { B[i] -= A[i * n + j] * B[j]; }
-            B[i] /= A[i * n + i];  // Divide by U's diagonal element
+        const SOLVER_VEC<Scalar> abs2_gamma = gamma_r * gamma_r + gamma_i * gamma_i;
+        const SOLVER_VEC<Scalar> abs_gamma  = solver_vec_sqrt<Scalar>(abs2_gamma);
+        condition_bound *= (one + abs_gamma) / (one - abs_gamma);
```
Comment (lgarrison) [inline src/nifty_ls/chi2_helpers.hpp:319]: Is there any concern about undefined behavior when performing operations that may divide by zero with `-ffast-math`? If it's too much trouble, I think we can probably re-enable `-ffinite-math`.

File: src/nifty_ls/chi2_helpers.hpp [hunk 5]
```diff
@@ -210,89 +210,165 @@ void compute_t_raw(
...
 }
 
 template <typename Scalar>
 void process_chi2_outputs_raw(
-   Scalar *power,       // (Nbatch, Nf)
-   const Scalar *Sw,    // input, (Nbatch, nSW, Nf)
-   const Scalar *Cw,    // input, (Nbatch, nSW, Nf)
-   const Scalar *Syw,   // input, (Nbatch, nSY, Nf)
-   const Scalar *Cyw,   // input, (Nbatch, nSY, Nf)
-   const Scalar *norm,  // input, (Nbatch, 1)
```
Comment (lgarrison) [inline src/nifty_ls/chi2_helpers.hpp:295]: Can you restore these comments?

File: src/nifty_ls/chi2_helpers.hpp [hunk 6]
```diff
@@ -210,89 +210,165 @@ void compute_t_raw(
...
+    union {
+        float f;
+        uint32_t u;
+    } bits = {value};
+    return (bits.u & UINT32_C(0x7fffffff)) >= UINT32_C(0x7f800000);
+}
+
+template <typename Scalar>
+bool solver_condition_is_singular(const Scalar condition_bound) {
+    const Scalar threshold = static_cast<Scalar>(1e11);
```
Comment (lgarrison) [inline src/nifty_ls/chi2_helpers.hpp:249]: Should this value be lower if `Scalar` is `float`? Alternatively, we could just ignore the `float` path if you think no one working on periodograms is going to use it.

**Review by lgarrison — CHANGES_REQUESTED** [review]
Thanks, I think this looks very nice! Most of my comments are cosmetic. I did some quick benchmarks locally and it did seem to give a nice speedup (2x in some cases).

Speaking of cosmetic: would you mind breaking out the Toeplitz solver into its own file? It can even be `#include`-d directly into `chi2_helpers.hpp`; just trying to keep the generic matrix solver logic separate from the Lomb-Scargle logic.

I also added some tests that exercise the chi2 paths a little more thoroughly. Can you take a look and let me know if you agree with them?

Finally, I think we need to update the Python API docstrings to explain the new behavior of returning NANs at ill-conditioned frequencies, reserving `LinAlgError` for fully NAN results. (And I'm not even sure we need `LinAlgError` at all any more; I'm open to thoughts on this.)

**Comment by kkrutkowski** [comment]
> new behavior of returning NANs at ill-conditioned frequencies, reserving `LinAlgError` for fully NAN results. (And I'm not even sure we need `LinAlgError` at all any more; I'm open to thoughts on this.)

Thank you very much, I'll try to do that soon. Also the NaN for now is only a temporary solution and I am not really sure what the nifty-ls really aims for. There are some alternatives like the fallback SVD solve returning rank-deficient solutions instead (which is really slow, but returns something that makes sense). 

Also there are quite some other potential choices waiting to be made - that's why that thing here is marked only as a draft. I'll try to 'fix' everything pretty soon, although I don't think it's anywhere close to being 'production-ready' yet.

File: src/nifty_ls/chi2_helpers.hpp (hunk 6 above)
Comment (kkrutkowski) [reply src/nifty_ls/chi2_helpers.hpp:249]: This one should be much lower for float, depending on the FinuFFT-backend.

As for float vs double path… Well… I am using almost exclusively single precision path on my own data, but FiNuFFT here is the problem. It relies on 'brute-force' gridding, which for N target frequencies causes O(N) error propagation instead of O(sqrt(log(N))) achievable through full mantissa preservation with additional trig identities. As such the condition number ceiling for float should get really low and probably depend on the grid length (Something like 1e7/N probably?).

Also I've added the comparison of FiNUFFT vs 'exact gridding' NuFFT implementation in my reference repo some time ago (here is the link if you are willing to look into possible fixes/patches: https://github.com/kkrutkowski/Toeplitz-LS/tree/main/examples/2_FINUFFT ) - I think the 'twiddle ladder' approach could be used along with FiNUFFT to significantly reduce error propagation and make single precision more usable, but that one would require switching to guru interface. 

In its current form I don't think the single precision is usable in most cases, but that's a result of FiNUFFT's quirk without any work-arounds and not a direct precision issue.

**Comment by lgarrison** [comment]
Sure, no rush. I think we're just aiming for nifty-ls to be fast and accurate in most common cases, so I'm not too worried if this PR doesn't implement the most optimal solution for every corner case. So I'm happy to defer things like SVD fallback or "twiddle ladder" improvements to future work.

Interesting that the new test passes on Linux but not MacOS, not sure what to make of that.

**Comment by kkrutkowski** [comment]
> Interesting that the new test passes on Linux but not MacOS, not sure what to make of that.

I've just done some testing regarding accuracy, and I think I might have been overly optimistic when trying to determine the conditionality threshold - lowering it to old 1e9 or somewhere around might fix the issue.

I've compared typical R^2 errors for a 1000-measurement grid of Nifty-LS against both of my implementations as well, as new AstroPy FastChi2 (part of 8.0 release candidate), (which, by the way, happens to be more accurate than the Chi2 currently used in the tests) - looks like the margin of error I left there is likely not enough here. With ~ > 1e12 expected error I think that the numerical errors of FiNUFFT might just made f=0 grid point numerically non-singular, which *accidentally* did not happened on Linux due to different optimization pathways. Or at least that one seems most reasonable to me, now, than I looked into it in a little bit more detail.

>=========================== short test summary info ============================
>FAILED tests/test_ls.py::test_chi2_singular_returns_nan[finufft_chi2-2-True] - AssertionError: expected NaN at the >singular f=0 bin
>assert np.False_
> +  where np.False_ = <ufunc 'isnan'>(np.float64(4.2489443180145374e-10))
> +    where <ufunc 'isnan'> = np.isnan
>FAILED tests/test_ls.py::test_chi2_singular_returns_nan[finufft_chi2-2-False] - AssertionError: expected NaN at the >singular f=0 bin
>assert np.False_
> +  where np.False_ = <ufunc 'isnan'>(np.float64(1.8949998747842047e-10))
> +    where <ufunc 'isnan'> = np.isnan
>============= 2 failed, 86 passed, 37 skipped in 62.46s (0:01:02) ==============

degree1_synthetic_error_vs_grid

**Comment by lgarrison** [comment]
Sounds good, if you want to lower the condition threshold a bit, that's fine with me.

There's another finufft quirk that may be contributing to making the f=0 frequency non-singular, which is that finufft doesn't natively support r2c Fourier transforms, so we manually apply a phase-shift of N/2 to the non-uniform points so the -N/2 frequency becomes the 0 frequency. But as the finufft experts pointed out to me, this gives the low frequencies less numerical stability than one might naively expect. I think it's below the level that we care about, but also I didn't think that anyone cared about accuracy to better than 1e-12!
